### PR TITLE
authz: update authz JWT task

### PR DESCRIPTION
**Pull request by yangminzhu**
This PR updates the authz JWT task page that removes the duplicated contents (half of them) and adds clear explanation of the list-typed claims in JWT.

The update was previously reviewed by @adammil2000 in a google doc, thanks Adam for the review and all the comments.

### CODE REVIEW

**Comment by yangminzhu** [comment]
The lint seems checking the embedded command output for HTTP link, @sdake do you have any idea how to fix this? thanks.

```
- ./public/docs/tasks/security/authorization/authz-jwt/index.html
  *  External link https://raw.githubusercontent.com/istio/istio/release-1.5/secure.istio.io","scope":["scope1","scope2"],"sub":"testing failed: 404 No error
  *  External link https://raw.githubusercontent.com/istio/istio/release-1.5/secure.istio.io","sub":"testing failed: 404 No error
  *  External link https://raw.githubusercontent.com/istio/istio/release-1.5/secure.istio.io/testing failed: 404 No error
htmlproofer 3.12.0 | Error:  HTML-Proofer found 3 failures!
LINTING FAILED
Makefile.core.mk:85: recipe for target 'lint' failed
make: *** [lint] Error 1
```

**Comment by yangminzhu** [comment]
@adammil2000 @liminw Could you take a look? Thanks.

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+This task shows you how to set up an Istio authorization policy to enforce access
+based on the JSON Web Token (JWT). An Istio authorization policy supports both string typed
+and list-of-string typed JWT claims.
+
+## Before you begin
+
+Before you begin this task, perform the following actions:
+
+* Read the [Authorization](/docs/concepts/security/#authorization) and [Authentication](/docs/concepts/security/#authentication).
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: remove the word `the`

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+* Install Istio using the [Istio installation guide](/docs/setup/install/istioctl/).
+
+* Deploy two workloads: `httpbin` and `sleep`. Deploy these in one namespace,
+for example `foo`. Both workloads run with an Envoy proxy in front of each.
+Deploy the example namespace and workloads using these commands:
+
+    {{< text bash >}}
+    $ kubectl create ns foo
+    $ kubectl apply -f <(istioctl kube-inject -f samples/httpbin/httpbin.yaml) -n foo
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: you probably want `@samples/httpbin/htmpbin.yaml@`

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    $ kubectl apply -f <(istioctl kube-inject -f samples/sleep/sleep.yaml) -n foo
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: you want `@samples/sleep/sleep.yaml@`

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+* Verify that `sleep` successfully communicates with `httpbin` using this command:
+
+    {{< text bash >}}
+    $ kubectl exec $(kubectl get pod -l app=sleep -n foo -o jsonpath={.items..metadata.name}) -c sleep -n foo -- curl http://httpbin.foo:8000/ip -s -o /dev/null -w "%{http_code}\n"
+    200
+    {{< /text >}}
+
+{{< warning >}}
+If you don’t see the expected output, retry after a few seconds.
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: this could be improved by `If you don't see the expected output 200, ...`

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+Caching and propagation can cause a delay.
+{{< /warning >}}
+
+## Allow requests with valid JWT and list-typed claims
+
+1. The following command creates the `jwt-example` request authentication policy
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: on line 48 change `The policy` -> `This policy`

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+      selector:
+        matchLabels:
+          app: httpbin
+      jwtRules:
+      - issuer: "testing@secure.istio.io"
+        jwksUri: "{{< github_file >}}/security/tools/jwt/samples/jwks.json"
+    EOF
+    {{< /text >}}
+
+1. Verify that a request with an invalid JWT is denied.
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: numbered list instructions should typically end with a `:`. In this case, change `denied.` to -> `denied:`

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+{{< warning >}}
+If you don’t see the expected output, retry after a few seconds.
+Caching and propagation can cause a delay.
+{{< /warning >}}
+
+## Allow requests with valid JWT and list-typed claims
+
+1. The following command creates the `jwt-example` request authentication policy
+for the `httpbin` workload in the `foo` namespace. The policy allows the Envoy proxy in front of
+the `httpbin` workload to accept a JWT issued by `testing@secure.istio.io`.
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: conclude ordered and unordered instructions with `:`.

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+    {{< /text >}}
+
+1. Verify that a request with an invalid JWT is denied.
+
+    {{< text bash >}}
+    $ kubectl exec $(kubectl get pod -l app=sleep -n foo -o jsonpath={.items..metadata.name}) -c sleep -n foo -- curl "http://httpbin.foo:8000/headers" -s -o /dev/null -H "Authorization: Bearer invalidToken" -w "%{http_code}\n"
+    401
+    {{< /text >}}
+
+1. Verify that a request without a JWT is allowed because there is no authorization policy.
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: terminate with `:`.

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+      rules:
+      - from:
+        - source:
+           requestPrincipals: ["testing@secure.istio.io/testing@secure.istio.io"]
+    EOF
+    {{< /text >}}
+
+1. Get the JWT that sets the `iss` and `sub` keys to the same value, `testing@secure.istio.io`.
+This causes Istio to generate the attribute `requestPrincipal` with the value,
+`testing@secure.istio.io/testing@secure.istio.io`.
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: terminate list with `:`

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+      action: ALLOW
+      rules:
+      - from:
+        - source:
+           requestPrincipals: ["testing@secure.istio.io/testing@secure.istio.io"]
+    EOF
+    {{< /text >}}
+
+1. Get the JWT that sets the `iss` and `sub` keys to the same value, `testing@secure.istio.io`.
+This causes Istio to generate the attribute `requestPrincipal` with the value,
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: remove `,`

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+`testing@secure.istio.io/testing@secure.istio.io`.
+
+    {{< text syntax="bash" expandlinks="false" >}}
+    $ TOKEN=$(curl {{< github_file >}}/security/tools/jwt/samples/demo.jwt -s) && echo $TOKEN | cut -d '.' -f2 - | base64 -d -
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: Nice work sorting this out !

`{{< github_file >}}` usage is described here: https://istio.io/about/contribute/creating-and-editing-pages/#github

TLDR: use `@` when using a bash quote. use `{{< github_file>}}` to create a link to a file. If as in this example line 113 contains `@` that is not an expansion, use `expandlinks="false"`. This will create a problem in the infrastructure mentioned above.

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+        - source:
+           requestPrincipals: ["testing@secure.istio.io/testing@secure.istio.io"]
+    EOF
+    {{< /text >}}
+
+1. Get the JWT that sets the `iss` and `sub` keys to the same value, `testing@secure.istio.io`.
+This causes Istio to generate the attribute `requestPrincipal` with the value,
+`testing@secure.istio.io/testing@secure.istio.io`.
+
+    {{< text syntax="bash" expandlinks="false" >}}
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: use of `expandlinks="false"` is correct here.

Comment (sdake) [reply content/en/docs/tasks/security/authorization/authz-jwt/index.md]: The website infrastructure does not handle the condition of a link to a file expansion while in the same block two consecutive `@` is used for non-linked purposes. In this case, you want that link expansion to occur. This would be an area for improvement, although it seems like such an edge case, as to not be worth the trouble.

Comment (sdake) [reply content/en/docs/tasks/security/authorization/authz-jwt/index.md]: (no action needed).

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md (hunk 11 above)
Comment (sdake) [reply content/en/docs/tasks/security/authorization/authz-jwt/index.md]: (no action needed).

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+    $ TOKEN=$(curl {{< github_file >}}/security/tools/jwt/samples/demo.jwt -s) && echo $TOKEN | cut -d '.' -f2 - | base64 -d -
+    {"exp":4685989700,"foo":"bar","iat":1532389700,"iss":"testing@secure.istio.io","sub":"testing@secure.istio.io"}
+    {{< /text >}}
+
+1. Verify that a request with a valid JWT is allowed.
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: more `:`

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+
+    {{< text bash >}}
+    $ kubectl exec $(kubectl get pod -l app=sleep -n foo -o jsonpath={.items..metadata.name}) -c sleep -n foo -- curl "http://httpbin.foo:8000/headers" -s -o /dev/null -H "Authorization: Bearer $TOKEN" -w "%{http_code}\n"
+    200
+    {{< /text >}}
+
+1. Verify that a request without a JWT is denied.
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: more `:`

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+    {{< text bash >}}
+    $ kubectl exec $(kubectl get pod -l app=sleep -n foo -o jsonpath={.items..metadata.name}) -c sleep -n foo -- curl "http://httpbin.foo:8000/headers" -s -o /dev/null -w "%{http_code}\n"
+    403
+    {{< /text >}}
+
+1. The following command updates the `require-jwt` authorization policy to also require
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: remove the comma from `containing the value,`

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+    {{< /text >}}
+
+1. Get the JWT that sets the `groups` claim to a list of strings: `group1` and `group2`.
+
+    {{< text syntax="bash" expandlinks="false" >}}
+    $ TOKEN_GROUP=$(curl {{< github_file >}}/security/tools/jwt/samples/groups-scope.jwt -s) && echo $TOKEN_GROUP | cut -d '.' -f2 - | base64 -d -
+    {"exp":3537391104,"groups":["group1","group2"],"iat":1537391104,"iss":"testing@secure.istio.io","scope":["scope1","scope2"],"sub":"testing@secure.istio.io"}
+    {{< /text >}}
+
+1. Verify that a request with the JWT that includes `group1` in the `groups` claim is allowed.
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: `allowed.` to `allowed:`

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+1. Verify that a request with a JWT, which doesn’t have the `groups` claim is rejected.
+
+    {{< text bash >}}
+    $ kubectl exec $(kubectl get pod -l app=sleep -n foo -o jsonpath={.items..metadata.name}) -c sleep -n foo -- curl "http://httpbin.foo:8000/headers" -s -o /dev/null -H "Authorization: Bearer $TOKEN" -w "%{http_code}\n"
+    403
+    {{< /text >}}
+
+## Clean up
+
+1. Remove the namespace `foo` from your configuration:
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: remove `from your configuration`

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 18]
```diff
@@ -0,0 +1,182 @@
+---
+title: Authorization for JWT
+description: Shows how to set up access control with JWT in Istio.
```
Comment (sdake) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: `Shows how`->`How`

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+based on the JSON Web Token (JWT). An Istio authorization policy supports both string typed
+and list-of-string typed JWT claims.
+
+## Before you begin
+
+Before you begin this task, perform the following actions:
+
+* Read the [Authorization](/docs/concepts/security/#authorization) and [Authentication](/docs/concepts/security/#authentication).
+
+* Install Istio using the [Istio installation guide](/docs/setup/install/istioctl/).
```
Comment (liminw) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: Remove "the".

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md (hunk 7 above)
Comment (liminw) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]: The policy for `httpbin` workload accepts a JWT ...

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md (hunk 4 above)
Comment (yangminzhu) [reply content/en/docs/tasks/security/authorization/authz-jwt/index.md]: This is a general warning for all commands in the task, not specific to the `200` one above.

**Comment by yangminzhu** [comment]
@liminw @sdake Thanks for the review, comments addressed, PTAL.

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+description: How to set up access control with JWT in Istio.
+weight: 30
+keywords: [security,authorization,jwt,claim]
+aliases:
+    - /docs/tasks/security/rbac-groups/
+    - /docs/tasks/security/authorization/rbac-groups/
+---
+
+This task shows you how to set up an Istio authorization policy to enforce access
+based on the JSON Web Token (JWT). An Istio authorization policy supports both string typed
```
Comment (frankbu) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]:
```suggestion
based on a JSON Web Token (JWT). An Istio authorization policy supports both string typed
```

File: content/en/docs/tasks/security/authorization/authz-jwt/index.md [hunk 21]
```diff
@@ -0,0 +1,181 @@
+---
+title: Authorization for JWT
```
Comment (frankbu) [inline content/en/docs/tasks/security/authorization/authz-jwt/index.md]:
```suggestion
title: Authorization with JWT
```
